compiler/src/dmd/statementsem.d

```diff
@@ -2818,7 +2818,10 @@ package (dmd) extern (C++) final class StatementSemanticVisitor : Visitor
             rs.exp = rs.exp.expressionSemantic(sc);
             rs.exp = rs.exp.arrayFuncConv(sc);
             // If we're returning by ref, allow the expression to be `shared`
-            const returnSharedRef = (tf.isref && (fd.inferRetType || tret.isShared()));
+
+            // https://issues.dlang.org/show_bug.cgi?id=23709
+            // Classes are reference types, therefore we can return shared instances
```

Comment:
what about AAs?

Reply:
Well, there might be other situations such as arrays to shared elements. I don't know the entire list, I'm simply fixing the reported issue.

We can add them once they are reported. It makes reviewing much easier.

Reply:
Does this work for interfaces as well?

Reply:
Yes.  Interfaces are also typed as `TypeClass`.